### DISCUSSION TITLE
Is it possible to add specific properties to gradle.properties file from config.xml?

### DISCUSSION THREAD

**Discussion by PeterHdd** [apache/cordova#482] [Q&A] [answered]
Hello Cordova android community,

First thank you for all the effort in the android platform! I have a question, is there a way to add properties to the `gradle.properties` file automatically? So basically when doing `cordova build android` I want a specific property to be added there so the build passes in the CI. I know for androidx we can just add this in the `config.xml`:

```
<preference name="AndroidXEnabled" value="true" />
```
But what if some other property needs to be added how would we do that?

For example if i wanna ignore jetifying the `jackson-core` library:

```
android.jetifier.ignorelist = jackson-core
```

Can this somehow be added in the `config.xml` or the `plugin.xml` and then it gets added to the `gradle.properties`? If so how can this be done?

**Comment by breautek** [comment, marked as answer]
There's nothing like `<edit-config>` or `<config-file>` (which only operates on XML files) for the gradle properties file.

But you can configure an `after_prepare` [hook](https://cordova.apache.org/docs/en/12.x/guide/appdev/hooks/index.html) to run a NodeJS script to manipulate the `gradle.properties`

The config would look something like:

```xml
<widget ...>
    ...
    <platform name="android">
        ...
        <hook type="after_prepare" src="scripts/configureJetifierIgnore.js" />
    </platform>
</widget>
```

And inside `scripts/configureJetifierIgnore.js`:

```js
const FS = require('fs');
const Path = require('path');

module.exports = function(context) {
    const propertiesPath = Path.resolve(process.cwd(), 'platforms/android/gradle.properties');
    
    let properties = FS.readFileSync(propertiesPath, {
        encoding: 'utf-8'
    });

    // Replace any lines that begin with android.jetifier.ignoreList
    properties = properties.replace(/^android\.jetifier\.ignorelist.+/m, '');

    // Append a new ignoreList directive on a new line:
    properties += '\r\nandroid.jetifier.ignoreList = jackson-core';

    // Finally write the file back out
    FS.writeFileSync(propertiesPath, properties);
};
```

Notes:
1. Code is untested, I don't remember what the actual current working directory is in hooks, but I assume it's the project root directory, where you normally run your `cordova` commands. The example code might require tweaks to work properly.
2. `after_prepare` hook is the hook after cordova sets up the native project. By this time all files that cordova manipulates/updates should be written. So any manipulations done by the hook should persist for the build step which starts just after the `prepare` step.
3. I do not remember if `gradle.properties` is one of the files that is consistently wiped, so you may have to check if `android.jetifier.ignoreList` key already exist in properties and either remove it (like the script does) or parse to update it.
4. The context object isn't well documented. What it contains will vary on hook type, platform, and CLI options passed. You can console.log it to see if there are anything that is helpful.

**Reply by PeterHdd** [reply]
Thank you, worked perfectly!
